vital/util/span.h

```diff
@@ -0,0 +1,153 @@
+#include <type_traits>
+
+#include <cmath>
+
+namespace kwiver {
+
+namespace vital {
+
+// ----------------------------------------------------------------------------
+/// Represents a numerical span with an upper and lower bound.
+template < class T >
+class VITAL_UTIL_EXPORT span
+{
+public:
+  span( T lower, T upper )
+    : m_lower{ std::min( lower, upper ) },
+      m_upper{ std::max( lower, upper ) }
+  {
+    assert_not_nan( lower );
+    assert_not_nan( upper );
+  }
+
+  /// Return lower bound.
+  T
+  lower() const
+  {
+    return m_lower;
+  }
+
+  /// Return upper bound.
+  T
+  upper() const
+  {
+    return m_upper;
+  }
+
+  /// Ensure the lower bound it at least \p new_lower.
+  ///
+  /// \throws invalid_argument If \p new_lower is greater than the current
+  /// upper bound.
+  void
+  truncate_lower( T new_lower )
+  {
+    assert_not_nan( new_lower );
+    if( new_lower > m_upper )
+    {
+      throw std::invalid_argument(
+              "span.truncate_lower(): "
+              "new_lower cannot be greater than current upper" );
+    }
+    m_lower = std::max( m_lower, new_lower );
+  }
+
+  /// Ensure the upper bound it at most \p new_upper.
+  ///
+  /// \throws invalid_argument If \p new_upper is less than the current lower
+  /// bound.
+  void
+  truncate_upper( T new_upper )
+  {
+    assert_not_nan( new_upper );
+    if( new_upper < m_lower )
+    {
+      throw std::invalid_argument(
+              "span.truncate_upper(): "
+              "new_upper cannot be less than current lower" );
+    }
+    m_upper = std::min( m_upper, new_upper );
+  }
+
+  /// Ensure the span contains \p value.
+  void
+  encompass( T value )
+  {
+    assert_not_nan( value );
+    m_lower = std::min( m_lower, value );
+    m_upper = std::max( m_upper, value );
+  }
+
+  /// Return \c true if \p value is within in this half-open span.
+  bool
+  contains( T value ) const
+  {
+    return m_lower <= value && value < m_upper;
+  }
+
+  /// Return \c true if \p value is within in this span.
+  ///
+  /// \param inclusive_lower Whether to return \c true if \p value is equal to
+  /// the lower bound.
+  /// \param inclusive_upper Whether to return \c true if \p value is equal to
+  /// the upper bound.
+  bool
+  contains( T value, bool inclusive_lower, bool inclusive_upper ) const
+  {
+    return ( m_lower < value || ( inclusive_lower && m_lower == value ) ) &&
+           ( value < m_upper || ( inclusive_upper && m_upper == value ) );
+  }
+
+private:
+  template < class S,
+             typename std::enable_if<
+               std::is_floating_point< S >::value, bool >::type = true >
+  static void
+  assert_not_nan( S value )
+  {
+    if( std::isnan( value ) )
+    {
+      throw std::invalid_argument( "span: cannot accept NaN value" );
+    }
+  }
+
+  template < class S,
+             typename std::enable_if<
+               !std::is_floating_point< S >::value, bool >::type = true >
+  static void assert_not_nan( S value ) {}
```

Comment:
```suggestion
  static void
  assert_not_nan( T value )
  {
    if( std::isnan( value ) )
    {
      throw std::invalid_argument( "span: cannot accept NaN value" );
    }
  }
```

...[`std::isnan`](https://en.cppreference.com/w/cpp/numeric/math/isnan) is actually specified for integer types, and the compiler will optimize away both the call and the enclosed `throw`, since it is of course always false for integer types. ([Demonstration](https://godbolt.org/z/hn6nY8xeP))

Reply:
Unfortunately, [Windows is not cooperating](https://godbolt.org/z/b3E5b5bzE)...

This causes our CI build for Windows to fail too, so I have reverted back to the original version of `assert_not_nan`.

Reply:
To make this work on Windows _and_ gcc, you could use `std::isnan( (float)value )`.  The Windows problem is coming from the fact that its `std::isnan` apparently doesn't have an implementation for `long long` and it isn't smart enough to figure out a conversion.  It's not elegant, but it works.  Since the point is only to determine if the value is `NaN` or not, any loss of precision in casting shouldn't matter.